File: src/main/java/leetcode/ContainerWithMostWater.java
package leetcode;

/**
 * Given n non-negative integers a1, a2, ..., an , where each represents a point
 * at coordinate (i, ai). n vertical lines are drawn such that the two endpoints
 * of line i is at (i, ai) and (i, 0). Find two lines, which together with
 * x-axis forms a container, such that the container contains the most water.
 * 
 * Note: You may not slant the container and n is at least 2.
 *
 */
class ContainerWithMostWater {
	/**
	 * Return maximum area.
	 *
	 * @param height an array of coordinates
	 * @return maximum area
	 */
	public static int maxArea(int[] height) {
		int maxArea = 0;
		int area = 0;
		for (int i = 0; i < height.length; i++) {
			for (int j = i; j < height.length; j++) {
				if (height[i] > height[j]) {
					area = height[j] * Math.abs(i - j);
				} else {
					area = height[i] * Math.abs(i - j);
				}

				if (area > maxArea) {
					maxArea = area;
				}
			}
		}
		return maxArea;
	}

}

File: python/src/container_with_most_water.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        """You are given an integer array height of length n. There are n vertical lines drawn such that the two
        endpoints of the ith line are (i, 0) and (i, height[i]). Find two lines that together with the x-axis form a
        container, such that the container contains the most water. Return the maximum amount of water a container
        can store. Notice that you may not slant the container.

        :param height: The height of the container.

        The time complexity is O(N) where N is the length of the input array. Faster than 92.32% of solutions.

        The space complexity is O(1) as no additional data structures are used. Less memory than 53.94% of solutions.

        """
        l, r = 0, len(height) - 1
        maxArea = -1

        while l < r:
            current_width = r - l
            current_height = min(height[l], height[r])
            maxArea = max(current_width * current_height, maxArea)
            if height[l] > height[r]:
                r -= 1
            else:
                l += 1

        return maxArea
